recipes/libelfin/all/conanfile.py

```diff
@@ -1,64 +1,78 @@
-    default_options = {"shared": False, "fPIC": True}
-
-    exports_sources = "CMakeLists.txt", "patches/*"
-    generators = "cmake"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
 
-    _cmake = None
-    _source_subfolder = "source_subfolder"
+    def export_sources(self):
+        copy(self, "CMakeLists.txt", src=self.recipe_folder, dst=self.export_sources_folder)
+        export_conandata_patches(self)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
     def configure(self):
-        if self.settings.compiler == "Visual Studio":
-            raise ConanInvalidConfiguration("libelfin doesn't support compiler: {} on OS: {}.".
-                                            format(self.settings.compiler, self.settings.os))
         if self.options.shared:
-            del self.options.fPIC
-        if self.settings.compiler.cppstd:
-            tools.check_min_cppstd(self, "11")
+            self.options.rm_safe("fPIC")
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def validate(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            check_min_cppstd(self, 11)
+        if is_msvc(self):
+            raise ConanInvalidConfiguration(f"libelfin doesn't support compiler: {self.settings.compiler}.")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version])
-        extracted_dir = self.name + "-" + self.version
-        os.rename(extracted_dir, self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["libelfin_VERSION"] = self.version
+        tc.generate()
 
-    def _configure_cmake(self):
-        if self._cmake:
-            return self._cmake
-        self._cmake = CMake(self)
-        self._cmake.configure()
-        return self._cmake
+    def _patch_sources(self):
+        apply_conandata_patches(self)
+        copy(self, "CMakeLists.txt", src=self.export_sources_folder, dst=self.source_folder)
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        cmake = self._configure_cmake()
+        self._patch_sources()
+        cmake = CMake(self)
+        cmake.configure()
         cmake.build()
 
     def package(self):
-        cmake = self._configure_cmake()
+        cmake = CMake(self)
         cmake.install()
 
     def package_info(self):
-        self.cpp_info.components["libelf++"].names["pkg_config"] = "libelf++"
+        self.cpp_info.components["libelf++"].set_property("pkg_config_name", "libelf++")
         self.cpp_info.components["libelf++"].libs = ["elf++"]
-        self.cpp_info.components["libdwarf++"].names["pkg_config"] = "libdwarf++"
```

Comment:
Keeping those names at the bottom is safer because we're still supporting Conan 1.x. Something like:

```python
    def package_info(self):
        self.cpp_info.components["libelf++"].set_property("pkg_config_name", "libelf++")
        self.cpp_info.components["libelf++"].libs = ["elf++"]
        self.cpp_info.components["libdwarf++"].set_property("pkg_config_name", "libdwarf++")
        self.cpp_info.components["libdwarf++"].libs = ["dwarf++"]
        self.cpp_info.components["libdwarf++"].requires = ["libelf++"]

        # TODO: Legacy, to be removed on Conan 2.0
        self.cpp_info.components["libelf++"].names["pkg_config"] = "libelf++"
        self.cpp_info.components["libdwarf++"].names["pkg_config"] = "libdwarf++"
```